src/util.cpp

```diff
@@ -9,20 +12,43 @@ std::vector<std::string> get_lines(std::filesystem::path const& filename) {
 	if (auto file = std::ifstream(filename)) {
 		while (std::getline(file, temp_line)) { ret.push_back(temp_line); }
 	} else {
-		std::cout << "File failed to open\n";
+		if (pn::pini::on_msg_t) {
+			std::stringstream str;
+			str << "File \"" << filename << "\"failed to open \n";
+			(*pn::pini::on_msg_t)(str.str(), pn::pini::severity::error);
+		}
 	}
 	return ret;
 }
 
 std::unordered_map<std::string, std::string> insert_pairs(std::vector<std::string> file_lines) {
 	std::unordered_map<std::string, std::string> key_value_pairs;
-
+	std::size_t line_number = 0;
 	for (const auto& str : file_lines) {
+		++line_number;
 		if (auto delimeter_pos = str.find('='); delimeter_pos != std::string::npos) {
-			key_value_pairs.insert({str.substr(0, delimeter_pos), str.substr(delimeter_pos + 1)});
+			std::string const key = trim_whitespace(str.substr(0, delimeter_pos));
+			if (key.empty() && pn::pini::on_msg_t) {
+				std::stringstream str;
+				str << "Line number " << line_number << " is not valid because the key is empty.\n";
+				(*pn::pini::on_msg_t)(str.str(), pn::pini::severity::warn);
+			} else {
+				std::string const value = trim_whitespace(str.substr(delimeter_pos + 1));
+				if (key[0] != '#') { key_value_pairs.insert({key, value}); }
+			}
 		}
 	}
 	return key_value_pairs;
 }
 
+std::string trim_whitespace(std::string const& str) {
```

Comment:
Again, these can be string views, you aren't modifying any chars, and the returned string is a contiguous substring of the passed one